### PR TITLE
dev,bazel: miscellaneous improvements

### PR DESCRIPTION
**Pull request by irfansharif**
**build,bazel: avoid stamping builds with commit sha**

When building development binaries, we already skip tagging each build
with the UTC time to avoid busting the build cache. We could go further
and not tag crdb binaries with the current SHA/last git tag with the
knowledge that we’re not shipping binaries from local machines. This
shaves off around ~30s of go link time on my machine when switching
branches, regardless of cache hit rate. It also shaves a 30s delay
whenever a commit is amended/rebased but the source is unaffected (when
rewording commits for e.g., or during interactive rebase).

**dev: improve help message**

CRDB-specific configs should probably be placed in the gitignored
.bazelrc.user. Using the top-level rc file mucks with bazel builds for
other repositories.

**bazel: avoid writing to bazel-out**

In #74771 we put a regular file in bazel-out to prevent bazel from doing
it automatically (this was to prevent tools from tripping over when
finding extra go files). Bazel exposes a top-level flag we can use
instead.

### CODE REVIEW

File: pkg/cmd/dev/doctor.go [hunk 1]
```diff
@@ -85,7 +85,7 @@ Please perform the following steps:
 	if !passedStampTest {
 		success = false
 		log.Printf(`Your machine is not configured to "stamp" your built executables.
-Please add one of the following to your ~/.bazelrc:`)
+Please add one of the following to your <path to cockroachdb checkout>/.bazelrc.user:`)
```
Comment (rickystewart) [inline pkg/cmd/dev/doctor.go:88]: Nit: consider just calling `d.getWorkspace()` to get the path to the checkout.

File: .bazelrc [hunk 2]
```diff
@@ -1,12 +1,17 @@
 try-import %workspace%/.bazelrc.user
 
-build --symlink_prefix=_bazel/ --ui_event_filters=-DEBUG --define gotags=bazel,gss --experimental_proto_descriptor_sets_include_source_info --incompatible_strict_action_env --incompatible_enable_cc_toolchain_resolution
-test --config=test --experimental_ui_max_stdouterr_bytes=10485760
+build --define gotags=bazel,gss
+build --experimental_proto_descriptor_sets_include_source_info
+build --incompatible_strict_action_env --incompatible_enable_cc_toolchain_resolution
+build --symlink_prefix=_bazel/ --experimental_no_product_name_out_symlink
+test --config=test --experimental_ui_max_stdouterr_bytes=10485760 --experimental_no_product_name_out_symlink
```
Comment (rickystewart) [inline .bazelrc:7]: Nit: `test` inherits from `build`, so `--experimental_no_product_name_out_symlink` is implied.

File: .bazelrc [hunk 3]
```diff
@@ -1,12 +1,17 @@
...
 build:with_ui --define cockroach_with_ui=y
 build:test --define crdb_test=y
 build:race --@io_bazel_rules_go//go/config:race --test_env=GORACE=halt_on_error=1 --test_sharding_strategy=disabled
 test:test --test_env=TZ=
 test:race --test_timeout=1200,6000,18000,72000
+
+build --ui_event_filters=-DEBUG
 query --ui_event_filters=-DEBUG
```
Comment (rickystewart) [inline .bazelrc:15]: Not sure whether `query` needs `--experimental_no_product_name_out_symlink` as well?

**Review by rickystewart — APPROVED** [review]

File: .bazelrc (hunk 3 above)
Comment (irfansharif) [reply .bazelrc:15]:
```
INFO: Reading rc options for 'query' from /Users/irfansharif/Software/src/github.com/cockroachdb/cockroach/.bazelrc:
  'query' options: --ui_event_filters=-DEBUG --experimental_no_product_name_out_symlink
ERROR: --experimental_no_product_name_out_symlink :: Unrecognized option: --experimental_no_product_name_out_symlink
```
Nah.

**Review by rickystewart — APPROVED** [review]

**Comment by irfansharif** [comment]
Thanks!

bors r+

**Pull request merged**